druid-shell/src/platform/x11/application.rs

```diff
@@ -404,4 +466,57 @@ impl Application {
         log::warn!("Application::get_locale is currently unimplemented for X11 platforms. (defaulting to en-US)");
         "en-US".into()
     }
+
+    pub(crate) fn idle_pipe(&self) -> RawFd {
+        self.idle_write
+    }
+}
+
+/// Returns when there is an event ready to read from `conn`, or there is data ready to read from
+/// `idle` and the `timeout` has passed.
+// This was taken, with minor modifications, from the xclock_utc example in the x11rb crate.
+// https://raw.githubusercontent.com/psychon/x11rb/a6bd1453fd8e931394b9b1f2185fad48b7cca5fe/examples/xclock_utc.rs
+fn poll_with_timeout(conn: &Rc<XCBConnection>, idle: RawFd, timeout: Instant) -> Result<(), Error> {
+    use nix::poll::{poll, PollFd, PollFlags};
+    use std::os::raw::c_int;
+    use std::os::unix::io::AsRawFd;
+
+    let fd = conn.as_raw_fd();
+    let mut poll_fds = [
+        PollFd::new(fd, PollFlags::POLLIN),
+        PollFd::new(idle, PollFlags::POLLIN),
+    ];
+
+    // We start with no timeout in the poll call. If we get something from the idle handler, we'll
+    // start setting one.
```

Comment:
This comment makes no sense to me and seems to be related to my comment below about the 100% CPU usage.

No new idle handler can appear while this code is running. Or are you working with threads? If so, this whole approach would break down (because another thread could cause x11rb to read in events, so events would end up in the internal queue, but the socket would not be readable, so `poll` would not return).

Later edit: Actually... this whole self-pipe thread looks like something that is needed only when threads are involved. Without threads, you cannot be blocking in `poll()` while at the same time deciding to need to wake up. With threads, your use of x11rb seems racy (I am open to suggestions on how to improve x11rb here, but I do not see any and especially `XCBConnection` is based on libxcb and cannot be modified here anyway).